laurat007/assignment1/Q1.java

```diff
@@ -32,102 +33,114 @@
 	
 	// number of pairs to be checked for being anagrams
 	static int N;
+        static Scanner sc;
 	
 	static boolean caseSensitiveAnagrams(String s1, String s2) {
-		System.out.println(s1 + " " + s2);
-		if(s1.length() != s2.length())
-			return false;
-		
-		// 58 = 122('z') - 65('A') + 1
-		int letters_occurrences[] = new int[58];
-		for(int i = 0 ; i < 58; i++)
-			letters_occurrences[i] = 0;
-		for(int i = 0; i < s1.length(); i++) {
-			letters_occurrences[122 - (int)s1.charAt(i)]++;
-		}
-		for(int i = 0; i < s1.length(); i++) {
-			letters_occurrences[122 - (int)s2.charAt(i)]--;
-		}
-		for(int i = 0 ; i < 57; i++)
-			if(letters_occurrences[i] != 0)
-				return false;
-		return true;
+            System.out.println(s1 + " " + s2);
+            if(s1.length() != s2.length())
+                return false;
+
+            // 58 = 122('z') - 65('A') + 1
+            int letters_occurrences[] = new int[58];
+            for(int i = 0 ; i < 58; i++)
+                letters_occurrences[i] = 0;
+            for(int i = 0; i < s1.length(); i++) {
+                letters_occurrences[122 - (int)s1.charAt(i)]++;
+            }
+            for(int i = 0; i < s1.length(); i++) {
+                letters_occurrences[122 - (int)s2.charAt(i)]--;
+            }
+            for(int i = 0 ; i < 57; i++)
+                if(letters_occurrences[i] != 0)
+                    return false;
+            return true;
 	}
+        
+        static boolean caseInsensitiveAnagrams(String s1, String s2) {
+            return caseSensitiveAnagrams(s1.toLowerCase(), s2.toLowerCase());
+        }
+        
+        static boolean caseInsensitiveSentenceAnagrams(ArrayList<String> s1, ArrayList<String> s2) {
+            if(s1.size() != s2.size()) {
+                return false;
+            }
+
+            // check if each word in first sentence
+            // matches any word in the second sentence
+            int exit = 0;
+            for(int j = 0; j < s1.size(); j++) {
+                exit = 0;
+                for(int k = 0; k < s2.size(); k++) {
+                    if(caseSensitiveAnagrams(s1.get(j).toLowerCase(),
```

Comment:
This is called O(N^2) times. For some cases of large sets of words that could have a significant run time. Can you think of changing your algorithm here to have linear running time?

Reply:
I changed it using some sorting algorithms, but I am not sure this is better as I still sort N words of length M, so it is not linear. Could you please give me an idea of what to do to make it better?